Initial raft code

plorenz opened the pull request.

andrewpmartinez commented on controller/agent.go:
```diff
@@ -51,3 +81,92 @@ func (self *Controller) agentOpSnapshotDb(c *bufio.ReadWriter) error {
+func (self *Controller) agentOpRaftRemove(m *channel.Message, ch channel.Channel) {
+	//id := string(m.Body)
+
+	// TODO: make this work like Join where we test if we're bootstrapped yet
+	//if err := self.raftController.HandleRemove(); err != nil {
+	//	return err
+	//}
+	// _, err := c.WriteString("success\n")
+	handler_common.SendOpResult(m, ch, "raft.remove", "no yet implemented", false)
```
todo?

plorenz replied:
Unless you think it's a bad idea, I'm not going to focus on having the raft management pieces fully baked at this stage. I want this pass focused on having the command infrastructure solid, then will do another pass (or passes, more likely) focused on solidifying the raft management pieces.

andrewpmartinez commented on controller/command/command.go:
```diff
@@ -0,0 +1,47 @@
+	"google.golang.org/protobuf/proto"
+)
+
+type Command interface {
+	Apply() error
+	Encode() ([]byte, error)
+}
+
+type Validatable interface {
+	Validate() error
+}
+
+type TypedMessage interface {
+	proto.Message
+	GetCommandType() int32
+}
```
Mysterious file. At least documenting the major interfaces would be nice.

plorenz replied:
Documented. Moved TypedMessage and Encode to the cmd_pb package

andrewpmartinez commented on controller/models/base_model.go:
```diff
@@ -44,6 +44,11 @@ type EntityRetriever interface {
 
 	GetStore() boltz.CrudStore
+
+	// GetEntityTypeId returns a unique id for the entity type. Some entities may share a storage type, such
+	//                 as fabric and edge services, and fabric and edge routers. However, they should have
+	//                 distinct entity type ids, so we can figure out which controller to route commands to
```
This extra loading space isn't something I've seen in the godoc standards.

https://go.dev/blog/godoc

andrewpmartinez commented on controller/network/command.go:
```diff
@@ -0,0 +1,130 @@
+	return decoder(msg.EntityData, updatedFields)
+}
+
+// CommandMsg is a TypedMessage which is also a pointer type.
+//
+// T is message type. We want to enforce that the TypeMessage implementation is a pointer type
+// so we can use new(T) to create instances of it
+type CommandMsg[T any] interface {
+	command.TypedMessage
+	*T
+}
+
+// decodableCommand is a Command which knows how to decode itself from the given message type
+//
+// T is the type of the command. We want to enforce that the command is a pointer type so we can
+// use new(T) to create new instances of it
+// M is the message type that the command can use to set its internals
+type decodableCommand[T any, M any] interface {
+	command.Command
+	Decode(n *Network, msg M) error
+	*T
+}
+
+// RegisterCommand register a decoder for the given command and message pair
+// MT is the message type (ex: cmd_pb.CreateServiceCommand)
+// CT is the command type (ex: CreateServiceCommand)
+// M is the CommandMsg/command.TypedMessage implementation (ex: *cmd_pb.CreateServiceCommand)
+// C is the decodableCommand/command.Command implementation (ex: *CreateServiceCommand)
+//
+// We only have both types specified so that we can enforce that each is a pointer type. If didn't
+// enforce that the instances were pointer types, we couldn't use new to instantiate new instances.
```
Love seeing this doc!

andrewpmartinez commented on controller/network/command.go:
```diff
@@ -0,0 +1,130 @@
+/*
```
Part of me wonders if there is a package in here for the command generics, command registry, etc. That could live on its own and be documented with an actual `package` level doc.

plorenz replied:
Working on consolidating generic (not generics) things in the command package.

andrewpmartinez commented on controller/raft/errors.go:
```diff
@@ -0,0 +1,25 @@
+	"reflect"
+)
+
+type Error interface {
```
Doc here please. Seeing an interface type called `Error` throws up so many questions.

plorenz replied:
Removing for now. Wasn't actually used anymore.

andrewpmartinez commented on controller/raft/fsm.go:
```diff
@@ -0,0 +1,146 @@
+				return err
+			}
+
+			logrus.Infof("apply log with type=%v", reflect.TypeOf(cmd))
```
can use `%T` here as an alternative to `TypeOf`

plorenz replied:
Fixed

andrewpmartinez commented on controller/raft/fsm.go:
```diff
@@ -0,0 +1,146 @@
+func (self *BoltDbFsm) Snapshot() (raft.FSMSnapshot, error) {
+	logrus.Info("creating snapshot")
+
+	file, err := ioutil.TempFile(self.dataDir, "raft-*.snapshot.db")
```
ooooh. Fancy. I didn't know that `TempFile` could take a glob.

andrewpmartinez commented on controller/raft/handlers.go:
```diff
@@ -0,0 +1,225 @@
+	ErrorCodeInvalidResult = 4
+)
+
+type Log interface {
```
Doc!

plorenz replied:
Was actually unused. removed

andrewpmartinez commented on controller/raft/handlers.go:
```diff
@@ -0,0 +1,225 @@
+}
+
+func sendErrorResponseCalculateType(m *channel.Message, ch channel.Channel, err error) {
+	// TODO: Handle more error types explicitly
```
Is there a reason to handle more explicitly? Is this just a maintenance block where we have to keep adding more types here?

plorenz replied:
Converted code to use ApiError consistenly when returning errors, so that way we only have api errors and any raft specific errors we may want to handle.

andrewpmartinez commented on controller/raft/hclog_adapter.go:
```diff
@@ -0,0 +1,231 @@
+
+import (
+	"fmt"
+	"github.com/hashicorp/go-hclog"
```
Is there a reason we need yet-another-logger?

plorenz replied:
hclog is what the hashicorp raft library uses. This is an adapter so the raft code logging will get converted to logrus and we can hopefully have consistent logging output

andrewpmartinez commented on controller/raft/mesh/mesh.go:
```diff
@@ -0,0 +1,285 @@
+	return self.addr
+}
+
+type Mesh interface {
```
Doc please!

plorenz replied:
fixed

andrewpmartinez commented on controller/raft/mesh/peerconn.go:
```diff
@@ -0,0 +1,159 @@
+	}
+}
+
+type raftPeerConn struct {
```
I have been mostly asking for doc on major interfaces, but this seems important. Consider doc?

plorenz replied:
Fixed

andrewpmartinez commented on controller/raft/raft.go:
```diff
@@ -0,0 +1,331 @@
+	return nil
+}
+
+// TODO: readOnly -> isVoter
```
Important or not?

plorenz replied:
fixed

andrewpmartinez commented on pb/cmd_pb/cmd.proto:
```diff
@@ -0,0 +1,66 @@
+}
+
+message CreateEntityCommand {
+  string entityType = 1;
+  bytes entityData = 2;
```
Does this need to be versioned? Or is versioning handled inside of `entityType`?

What happens if we have two different controller versions that expect two different byte interpretations of `entityData`?

plorenz replied:
I think we can handle versioning either in the message header (outside of the prototype) or in the entityData. We can also add a version field later in the protobuf without breaking backwards compatibility. Versioning still requires more thought.

andrewpmartinez reviewed (changes requested):
See comments. Great stuff overall though!

plorenz commented on controller/network/gen_test.go:
```diff
@@ -0,0 +1,50 @@
+	return decoder
+}
+
+func TestGen(t *testing.T) {
```
Some cruft from when I was putting together a test case for Goland generics false positive bug. Removing the whole file.

plorenz commented on controller/api/cors.go:
```diff
@@ -14,7 +18,8 @@ func WrapCorsHandler(innerHandler http.Handler) http.Handler {
 			"accept",
 			"authorization",
-			constants.ZitiSession,
+			// TODO: Not required for pure fabric. Is it worth having separate CorsHandlers for fabric and edge?
```
I'll leave it for now and remove the TODO

plorenz commented on controller/raft/handlers.go:
```diff
@@ -0,0 +1,225 @@
+// 2. Pre-submit validation failure
+// 3. Raft application failure
+
+// TODO: Use goroutines.Pool
```
Added goroutine pool

plorenz replied:
Added goroutine pool

plorenz commented on controller/raft/member.go:
```diff
@@ -0,0 +1,164 @@
+	Leader bool
+}
+
+type MemberModel interface {
```
fixed

plorenz commented on controller/raft/mesh/timeouts.go:
```diff
@@ -0,0 +1,42 @@
+	}
+}
+
+type deadline struct {
```
yeah, I thought I'd see if I either figured out a better way to do it, or if it was used in more places. Then it would either go away, or get moved to a common place.

plorenz commented on controller/raft/raft.go:
```diff
@@ -0,0 +1,331 @@
+	return result, nil
+}
+
+type Controller struct {
```
Documented

plorenz commented:
> I am guilty of this too, but w/ PRs if you can avoid force pushing, it makes it easier to come back into a review. Then rebase/squash after the PR is approved just before merging to keep the git graph nice.

Hmm, I purposely kept the follow up commits separate from the initial pr commit, so you could look at just the changes. I probably should have squashed those into a single commit, so you didn't have to look at multiple commits. 
I didn't think about that I'd be messing with githubs 'changes since last time' feature. Next time I'll try to keep the cleanup till the end.
